### PR TITLE
[RFC] [air] Torch off-the-shelf trainer

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This is a WIP version for an off-the-shelf torch trainer. Since many simple machine learning workflows follow the same pattern, we can make it easy for users to train simple models. Similar utilities exist e.g. in sklearn.

The off-the-shelf trainer is simple by design and is not aimed to be extended to more complicated use cases. We are not trying to build an alternative to pytorch lightning here. We just want to make the simplest training paradigms simple.

The pseudo code for the training loop we implement here is

```
        for x, y in training_dataset:
            output = model(x)
            loss = loss_fn(output, y)
            loss.backward()
            optimizer.step()

            maybe_validate_on(validation_dataset)
```

And a sample usage would be e.g.

```
trainer = TorchOffTheShelfTrainer(
    model_cls=NeuralNetwork,
    optimizer_cls=optim.SGD,
    optimizer_args={"lr": 0.05},
    scaling_config={"num_workers": 2},
    feature_columns=["x"],
    label_column="y",
    datasets={
        "train": lambda: get_dataset(train=True),
        "validate": lambda: get_dataset(train=False),
    },
)
trainer.fit()
```

This PR is a RFC to discuss if we want to have this utility at all, and if so to scope it. There are no unit tests and examples have not been added to CI, yet (but pass locally).

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/train/integrations/torch/torch_trainer.py [hunk 1]
```diff
@@ -109,7 +109,7 @@ def __init__(self):
                     super(NeuralNetwork, self).__init__()
                     self.layer1 = nn.Linear(input_size, layer_size)
                     self.relu = nn.ReLU()
-                    self.layer2 = nn.Linear(layer_size, output_size)
+                    self.layer2 = nn.Linear(layer_size, output_size)tr
```
Comment (maxpumperla) [inline python/ray/ml/train/integrations/torch/torch_trainer.py:112]: typo

File: python/ray/ml/examples/pytorch/off_the_shelf_fashion_mnist.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        )
+
+    def forward(self, x):
+        x = self.flatten(x)
+        return self.linear_relu_stack(x)
+
+
+# Initialize the off-the-shelf trainer and fit to the dataset
+
+trainer = TorchOffTheShelfTrainer(
```
Comment (maxpumperla) [inline python/ray/ml/examples/pytorch/off_the_shelf_fashion_mnist.py:44]: @krfricke thanks for this RFC! 

I think you could argue that something like this (not 100% about the exact specs yet) should be the default trainer users see. I'm not sure how often you really want to implement your own train loop in practice. This abstraction is certainly quicker to the punch for many use cases.

As you can imagine, I don't like the name. :D An idea would be to call this one `TorchTrainer` and the more complex underlying one something else.

Having said all that, I want to gauge the overall feeling about this from the rest of the pack first. As this diverges from the original design, I'm not sure we want to introduce another layer right now.

Comment (krfricke) [reply python/ray/ml/examples/pytorch/off_the_shelf_fashion_mnist.py:44]: Thanks for the feedback. Yes, this would be a higher level layer to make it very easy for users to start off. If we decide to do something like this, we should try to collect a lot of feedback to be sure we provide the right amount of customization while still keeping it very, very simple.

I'm always open to better naming ;-)

**Comment by Yard1** [comment]
@krfricke Thinking about it, maybe it would be better off as an example? That way we do not have to worry about not being customisable enough.

**Comment by amogkam** [comment]
Thanks @krfricke!  This looks really great, and I completely agree with @maxpumperla that this should actually be the primary API moving forward. Using off the shelf solution should be the default, but it should also be possible to implement your own training loop.

Want to throw out an alternative suggestion exploring the design space some more:
Instead of introducing a new Trainer, I think we can just provide the functions directly which users can use with the existing Trainers.

So something like (the name of the class is not great)

```python
SimpleTorchTrainLoopPerWorker:
    def __init__(self, model_cls, optimizer_cls, ...):
        self.model = model_cls()
        self.optimizer = optimizer_cls()
    def __call__(self):
        ....
```

I think this will allow for more composability. For example, you can imagine

```python
import horovod as hvd

SimpleHorovodTorchTrainLoopPerWorker(SimpleTorchTrainLoopPerWorker):
    def __init__(self, model_cls, optimizer_cls, ...):
        hvd.init()
        self.model = model_cls()
        self.optimizer = hvd.DistributedOptimizer(self.optimizer_cls(**self.optimizer_args))
```

**Comment by amogkam** [comment]
Basically we can create a bunch of these off the shelf training logic functions, without having to create a new Trainer for each one.

**Comment by richardliaw** [comment]
RE: primary api/being the default.

The utility of this depends on where are users are coming from right?

For example, if we think 90% of torch users are going to be migrating from existing code, then 90% of users ALREADY have a training loop in practice (or already have another higher level abstraction), and making this the default would **introduce cognitive friction** (as it would not be aiding the common case).

On the other hand, if we think 90% of AIR users are going to be new to pytorch, and this is their first exposure to pytorch, then certainly this would help.

Personally, I suspect there's a bit of maintainer bias here because we (as example writers) have to write our training loops from scratch.

In any case I think it doesn't hurt to merge as an example (and document it, marking as alpha stability) -- then leveraging our dogfooding process to see if most would find this useful, qualifying it to be the default.

**Comment by krfricke** [comment]
Ok sounds good, I'll update this PR. Any suggestion for naming? If this were to become the default, `TorchTrainer(TorchLoopTrainer)` would fit well (and the current `TorchTrainer` becomes the `TorchLoopTrainer`). But maybe we can stick with the current name for evaluation.

**Comment by richardliaw** [comment]
SimpleTorchTrainer?

**Comment by krfricke** [comment]
@amogkam I agree that a bunch of default train loops would serve this better here, there is no need to have an actual trainer. It could also lead to more confusion on which trainer to use.

Train loops also have the advantage that they can be inspected, copied, and adjusted to ones needs.

I'll close this PR for now, but we can maybe track this for documentation. cc @maxpumperla
